fdbserver/workloads/RemoveServersSafely.actor.cpp

```diff
@@ -401,11 +403,48 @@ struct RemoveServersSafelyWorkload : TestWorkload {
 
 		state std::vector<ISimulator::ProcessInfo*>	killProcArray;
 		state std::vector<AddressExclusion>	toKillArray;
+		state std::vector<AddressExclusion>	toKillMarkFailedArray;
 
 		std::copy(toKill.begin(), toKill.end(), std::back_inserter(toKillArray));
 		killProcArray = self->getProcesses(toKill);
+		if (safeKillSet) {
+			state int timeouts = 0;
+			loop {
+				state bool safe = false;
+				auto failSet = random_subset(toKillArray, deterministicRandom()->randomInt(0, toKillArray.size() + 1));
+				toKillMarkFailedArray.resize(failSet.size());
+				std::copy(failSet.begin(), failSet.end(), toKillMarkFailedArray.begin());
+				TraceEvent("RemoveAndKill", functionId)
+				    .detail("Step", "SafetyCheck")
+				    .detail("Exclusions", describe(toKillMarkFailedArray));
+				loop {
+					choose {
+						when(bool _safe = wait(checkSafeExclusions(cx, toKillMarkFailedArray))) {
+							safe = _safe;
+							break;
+						}
+						when(wait(delay(5.0))) {
+							TraceEvent("RemoveAndKill", functionId)
+							    .detail("Step", "SafetyCheckTimedOut")
+							    .detail("Exclusions", describe(toKillMarkFailedArray));
+						}
+					}
+					if (timeouts == self->maxSafetyCheckTimeouts) {
+						// Do not perform safety check, essentially simulating 'FORCE' option
+						TraceEvent("RemoveAndKill", functionId).detail("Step", "SafetyCheckLimitReached").detail("Timeouts", timeouts);
+						safe = true;
+						break;
+					}
+					timeouts++;
+				}
+				if (safe) break;
+			}
+		}
 
 		TraceEvent("RemoveAndKill", functionId).detail("Step", "Activate Server Exclusion").detail("KillAddrs", toKill.size()).detail("KillProcs", killProcArray.size()).detail("MissingProcs", toKill.size()!=killProcArray.size()).detail("ToKill", describe(toKill)).detail("Addresses", describe(toKillArray)).detail("ClusterAvailable", g_simulator.isAvailable());
+		if (safeKillSet) {
+			wait( excludeServers( cx, toKillMarkFailedArray, true ) );
+		}
 		wait( excludeServers( cx, toKillArray ) );
```

Comment:
I know this `wait( excludeServers( cx, toKillArray ) );` is not added in your PR.  However, since this PR adds new semantics to `exclude` command, we may want to test different scenarios: 1) some servers are marked as failed, some are just excluded, and some are both; 2) all servers are marked as failed; 3) all servers are marked as excluded; 4) all servers are marked as exclude while some are marked as failed as well.

Right now, this only tests the scenario 4).

Reply:
Functionally, I believe that `failed` should be a strict subset of `excluded`, so these scenarios are essentially the same. Anything marked as `failed` would be the same as if it were `excluded` (minus being able to include it again), so 1, 3, and 4 are no different from one another.

Reply:
Or more accurately, scenario 4 has MORE test coverage than 1 and 3, never less. Scenario 2 is also occasionally tested, depending on the outcome of the random subset size.